File: hexameter.core/src/test/java/org/codetome/hexameter/core/internal/impl/HexagonalGridImplTest.java
package org.codetome.hexameter.core.internal.impl;

import org.codetome.hexameter.core.api.*;
import org.codetome.hexameter.core.api.defaults.DefaultSatelliteData;
import org.codetome.hexameter.core.backport.Optional;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;
import static org.codetome.hexameter.core.api.CubeCoordinate.fromCoordinates;
import static org.codetome.hexameter.core.api.HexagonOrientation.POINTY_TOP;
import static org.codetome.hexameter.core.api.HexagonalGridLayout.RECTANGULAR;
import static org.junit.Assert.assertArrayEquals;

@SuppressWarnings("ALL")
public class HexagonalGridImplTest {

    private static final int RADIUS = 30;
    private static final int GRID_WIDTH = 10;
    private static final int GRID_HEIGHT = 10;
    private static final HexagonOrientation ORIENTATION = POINTY_TOP;
    private static final int GRID_X_FROM = 2;
    private static final int GRID_X_TO = 4;
    private static final int GRID_Z_FROM = 3;
    private static final int GRID_Z_TO = 5;
    private HexagonalGrid<DefaultSatelliteData> target;
    private HexagonalGridBuilder<DefaultSatelliteData> builder;

	@Before
    public void setUp() {
        builder = new HexagonalGridBuilder<DefaultSatelliteData>().setGridHeight(GRID_HEIGHT).setGridWidth(GRID_WIDTH).setRadius(RADIUS).setOrientation(ORIENTATION);
        target = builder.build();
    }

    @Test
    public void shouldReturnHexagonsInProperIterationOrderWhenGetHexagonsIsCalled() {
        final Collection<CubeCoordinate> expectedCoordinates = new ArrayList<>();
        final Collection<CubeCoordinate> actualCoordinates = new ArrayList<>();

        for (CubeCoordinate cubeCoordinate : builder.getGridLayoutStrategy().fetchGridCoordinates(builder)) {
            expectedCoordinates.add(cubeCoordinate);
        }
        for (Hexagon<DefaultSatelliteData> hexagon : target.getHexagons()) {
            actualCoordinates.add(hexagon.getCubeCoordinate());
        }

        assertArrayEquals(expectedCoordinates.toArray(), actualCoordinates.toArray());
    }

    @Test
    public void shouldReturnProperHexagonsWhenEachHexagonIsTestedInAGivenGrid() {
        final Iterable<Hexagon<DefaultSatelliteData>> hexagons = target.getHexagons();
        final Set<String> expectedCoordinates = new HashSet<>();
        for (int x = 0; x < GRID_WIDTH; x++) {
            for (int y = 0; y < GRID_HEIGHT; y++) {
                final int gridX = CoordinateConverter.convertOffsetCoordinatesToCubeX(x, y, ORIENTATION);
                final int gridZ = CoordinateConverter.convertOffsetCoordinatesToCubeZ(x, y, ORIENTATION);
                expectedCoordinates.add(gridX + "," + gridZ);
            }
        }
        final AtomicInteger count = new AtomicInteger();
        for (Hexagon<DefaultSatelliteData> hexagon : hexagons) {
            expectedCoordinates.remove(hexagon.getGridX() + "," + hexagon.getGridZ());
            count.incrementAndGet();
        }
        assertEquals(100, count.get());
        assertTrue(expectedCoordinates.isEmpty());
    }

    @Test
    public void shouldReturnProperHexagonsWhenGetHexagonsByAxialRangeIsCalled() {
        final Set<Hexagon> expected = new HashSet<>();

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 3)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 3)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 3)).get());

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 4)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 4)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 4)).get());

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 5)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 5)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 5)).get());

        final Iterable<Hexagon<DefaultSatelliteData>> actual = target.
                getHexagonsByCubeRange(fromCoordinates(GRID_X_FROM, GRID_Z_FROM), fromCoordinates(GRID_X_TO, GRID_Z_TO));
        final AtomicInteger count = new AtomicInteger();

        final List<Hexagon<DefaultSatelliteData>> actuals = new ArrayList<>();
        for (Hexagon<DefaultSatelliteData> hex : actual) {
            actuals.add(hex);
            count.incrementAndGet();
        }
        assertEquals(expected.size(), count.get());
        for (Hexagon<DefaultSatelliteData> hex : actuals) {
            expected.remove(hex);
        }
        assertTrue(expected.isEmpty());
    }

    @Test
    public void shouldReturnProperHexagonsWhenGetHexagonsByOffsetRangeIsCalled() {
        final Set<Hexagon> expected = new HashSet<>();

        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 3)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 3)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 3)).get());

        expected.add(target.getByCubeCoordinate(fromCoordinates(0, 4)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 4)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 4)).get());

        expected.add(target.getByCubeCoordinate(fromCoordinates(0, 5)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 5)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 5)).get());

        final Iterable<Hexagon<DefaultSatelliteData>> actual = target.getHexagonsByOffsetRange(GRID_X_FROM, GRID_X_TO, GRID_Z_FROM, GRID_Z_TO);
        final AtomicInteger count = new AtomicInteger();
        final List<Hexagon<DefaultSatelliteData>> actuals = new ArrayList<>();
        for (Hexagon<DefaultSatelliteData> hex : actual) {
            actuals.add(hex);
            count.incrementAndGet();
        }
        assertEquals(expected.size(), count.get());
        for (Hexagon<DefaultSatelliteData> hex : actuals) {
            expected.remove(hex);
        }
        assertTrue(expected.isEmpty());
    }

    @Test
    public void shouldContainCoordinateWhenContainsCoorinateIsCalledWithProperParameters() {
        final int gridX = 2;
        final int gridZ = 3;
        assertTrue(target.containsCubeCoordinate(fromCoordinates(gridX, gridZ)));
    }

    @Test
    public void shouldReturnHexagonWhenGetByGridCoordinateIsCalledWithProperCoordinates() {
        final int gridX = 2;
        final int gridZ = 3;
        final Optional<Hexagon<DefaultSatelliteData>> hex = target.getByCubeCoordinate(fromCoordinates(gridX, gridZ));
        assertTrue(hex.isPresent());
    }

    @Test()
    public void shouldBeEmptyWhenGetByGridCoordinateIsCalledWithInvalidCoordinates() {
        final int gridX = 20;
        final int gridZ = 30;
        Optional<Hexagon<DefaultSatelliteData>> result = target.getByCubeCoordinate(fromCoordinates(gridX, gridZ));
        Assert.assertFalse(result.isPresent());
    }

    @Test
    public void shouldReturnHexagonWhenCalledWithProperCoordinates() {
        final double x = 310;
        final double y = 255;
        final Hexagon hex = target.getByPixelCoordinate(x, y).get();
        assertTrue(hex.getGridX() == 3);
        assertTrue(hex.getGridZ() == 5);
    }

    @Test
    public void shouldReturnHexagonWhenCalledWithProperCoordinates2() {
        final double x = 300;
        final double y = 275;
        final Hexagon hex = target.getByPixelCoordinate(x, y).get();
        assertTrue(hex.getGridX() == 3);
        assertTrue(hex.getGridZ() == 5);
    }

    @Test
    public void shouldReturnHexagonWhenCalledWithProperCoordinates3() {
        final double x = 325;
        final double y = 275;
        final Hexagon hex = target.getByPixelCoordinate(x, y).get();
        assertTrue(hex.getGridX() == 3);
        assertTrue(hex.getGridZ() == 5);
    }

    @Test
    public void shouldReturnProperNeighborsOfHexagonWhenHexIsInMiddle() {
        final Hexagon hex = target.getByCubeCoordinate(fromCoordinates(3, 7)).get();
        final Set<Hexagon> expected = new HashSet<>();
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 6)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 6)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 7)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 8)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 8)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 7)).get());
        final Iterable<Hexagon<DefaultSatelliteData>> actual = target.getNeighborsOf(hex);
        assertEquals(expected, actual);
    }

    @Test
    public void shouldReturnProperNeighborsOfHexagonWhenHexIsOnTheEdge() {
        final Hexagon<DefaultSatelliteData> hex = target.getByCubeCoordinate(fromCoordinates(5, 9)).get();
        final Set<Hexagon<?>> expected = new HashSet<>();
        expected.add(target.getByCubeCoordinate(fromCoordinates(5, 8)).get());
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 9)).get());
        final Iterable<Hexagon<DefaultSatelliteData>> actual = target.getNeighborsOf(hex);
        assertEquals(expected, actual);
    }

    @Test
    public void shouldProperlyReturnGridLayoutWhenGetGridLayoutIsCalled() {
        Assert.assertEquals(RECTANGULAR, target.getGridData().getGridLayout());
    }

    @Test
    public void shouldProperlyReturnSharedHexagonDataWhenGetSharedHexagonDataIsCalled() {
        Assert.assertEquals(builder.getGridData().getHexagonHeight(), target.getGridData().getHexagonHeight(), 0);
        Assert.assertEquals(builder.getGridData().getHexagonWidth(), target.getGridData().getHexagonWidth(), 0);
        Assert.assertEquals(builder.getGridData().getRadius(), target.getGridData().getRadius(), 0);
        Assert.assertEquals(builder.getGridData().getOrientation(), target.getGridData().getOrientation());
    }

    @Test
    public void shouldProperlyReturnGridWidthWhenGetGridWidthIsCalled() {
        Assert.assertEquals(GRID_WIDTH, target.getGridData().getGridWidth());
    }

    @Test
    public void shouldProperlyReturnGridHeightWhenGetGridHeightIsCalled() {
        Assert.assertEquals(GRID_HEIGHT, target.getGridData().getGridHeight());
    }
}

File: mixite2.core/src/commonTest/kotlin/org/hexworks/mixite2/core/internal/impl/HexagonalGridImplTest.kt
package org.hexworks.mixite2.core.internal.impl

import org.hexworks.mixite2.core.api.*
import org.hexworks.mixite2.core.api.CubeCoordinate.Companion.fromCoordinates
import org.hexworks.mixite2.core.api.defaults.DefaultSatelliteData
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFalse

/**
 * This class does not actually run its own tests! This is a base class,
 * and the subclasses each run the inherited tests.
 */
abstract class HexagonalGridImplTest {

    internal lateinit var target: HexagonalGrid<DefaultSatelliteData>
    private lateinit var builder: HexagonalGridBuilder<DefaultSatelliteData>

    abstract fun getBuilder() : HexagonalGridBuilder<DefaultSatelliteData>

    @BeforeTest
    fun setUp() {
        builder = getBuilder()
        target = builder.build()
    }

    @Test
    fun shouldReturnHexagonsInProperIterationOrderWhenGetHexagonsIsCalled() {
        val expectedCoordinates = ArrayList<org.hexworks.mixite2.core.api.CubeCoordinate>()
        val actualCoordinates = ArrayList<org.hexworks.mixite2.core.api.CubeCoordinate>()

        for (cubeCoordinate in builder.gridLayoutStrategy.fetchGridCoordinates(builder)) {
            expectedCoordinates.add(cubeCoordinate)
        }
        for (hexagon in target.hexagons) {
            actualCoordinates.add(hexagon.cubeCoordinate)
        }

        assertEquals(expectedCoordinates, actualCoordinates)
    }

    @Test
    abstract fun shouldReturnProperHexagonsWhenEachHexagonIsTestedInAGivenGrid()

    @Test
    fun shouldReturnProperHexagonsWhenGetHexagonsByAxialRangeIsCalled() {
        val expected = HashSet<Hexagon<DefaultSatelliteData>>()

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 3)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 3)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 3)).get())

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 4)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 4)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 4)).get())

        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 5)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 5)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 5)).get())

        val actual = target.getHexagonsByCubeRange(fromCoordinates(GRID_X_FROM, GRID_Z_FROM), fromCoordinates(GRID_X_TO, GRID_Z_TO))
        var count = 0

        val actuals = ArrayList<Hexagon<DefaultSatelliteData>>()
        for (hex in actual) {
            actuals.add(hex)
            count++
        }
        assertEquals(expected.size, count)
        for (hex in actuals) {
            expected.remove(hex)
        }
        assertTrue(expected.isEmpty())
    }

    @Test
    fun shouldReturnProperHexagonsWhenGetHexagonsByOffsetRangeIsCalled() {
        val expected = HashSet<Hexagon<DefaultSatelliteData>>()

        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 3)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 3)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 3)).get())

        expected.add(target.getByCubeCoordinate(fromCoordinates(0, 4)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 4)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 4)).get())

        expected.add(target.getByCubeCoordinate(fromCoordinates(0, 5)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(1, 5)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 5)).get())

        val actual = target.getHexagonsByOffsetRange(GRID_X_FROM, GRID_X_TO, GRID_Z_FROM, GRID_Z_TO)
        var count = 0
        val actuals = ArrayList<Hexagon<DefaultSatelliteData>>()
        for (hex in actual) {
            actuals.add(hex)
            count++
        }
        assertEquals(expected.size, count)
        for (hex in actuals) {
            expected.remove(hex)
        }
        assertTrue(expected.isEmpty())
    }

    @Test
    fun shouldContainCoordinateWhenContainsCoorinateIsCalledWithProperParameters() {
        val gridX = 2
        val gridZ = 3
        assertTrue(target.containsCubeCoordinate(fromCoordinates(gridX, gridZ)))
    }

    @Test
    fun shouldReturnHexagonWhenGetByGridCoordinateIsCalledWithProperCoordinates() {
        val gridX = 2
        val gridZ = 3
        val hex = target.getByCubeCoordinate(fromCoordinates(gridX, gridZ))
        assertTrue(hex.isPresent)
    }

    @Test
    fun shouldBeEmptyWhenGetByGridCoordinateIsCalledWithInvalidCoordinates() {
        val gridX = 20
        val gridZ = 30
        val result = target.getByCubeCoordinate(fromCoordinates(gridX, gridZ))
        assertFalse(result.isPresent)
    }

    @Test
    fun shouldReturnHexagonWhenCalledWithProperCoordinates() {
        val x = 310.0
        val y = 255.0
        val hex = target.getByPixelCoordinate(x, y).get()
        assertEquals(hex.gridX, 3)
        assertEquals(hex.gridZ, 5)
    }

    @Test
    fun shouldReturnHexagonWhenCalledWithProperCoordinates2() {
        val x = 300.0
        val y = 275.0
        val hex = target.getByPixelCoordinate(x, y).get()
        assertEquals(hex.gridX, 3)
        assertEquals(hex.gridZ, 5)
    }

    @Test
    fun shouldReturnHexagonWhenCalledWithProperCoordinates3() {
        val x = 325.0
        val y = 275.0
        val hex = target.getByPixelCoordinate(x, y).get()
        assertEquals(hex.gridX, 3)
        assertEquals(hex.gridZ, 5)
    }

    @Test
    fun shouldReturnProperNeighborsOfHexagonWhenHexIsInMiddle() {
        val hex = target.getByCubeCoordinate(fromCoordinates(3, 7)).get()
        val expected = HashSet<Hexagon<DefaultSatelliteData>>()
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 6)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 6)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(4, 7)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(3, 8)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 8)).get())
        expected.add(target.getByCubeCoordinate(fromCoordinates(2, 7)).get())
        val actual = target.getNeighborsOf(hex)
        assertEquals(expected, actual)
    }

    @Test
    abstract fun shouldReturnProperNeighborsOfHexagonWhenHexIsOnTheEdge()

    @Test
    abstract fun shouldProperlyReturnGridLayoutWhenGetGridLayoutIsCalled()

    @Test
    fun shouldProperlyReturnSharedHexagonDataWhenGetSharedHexagonDataIsCalled() {
        assertEquals(builder.gridData.hexagonHeight, target.gridData.hexagonHeight)
        assertEquals(builder.gridData.hexagonWidth, target.gridData.hexagonWidth)
        assertEquals(builder.gridData.radius, target.gridData.radius)
        assertEquals(builder.gridData.orientation, target.gridData.orientation)
    }

    @Test
    fun shouldProperlyReturnGridWidthWhenGetGridWidthIsCalled() {
        assertEquals(GRID_WIDTH, target.gridData.gridWidth)
    }

    @Test
    fun shouldProperlyReturnGridHeightWhenGetGridHeightIsCalled() {
        assertEquals(GRID_HEIGHT, target.gridData.gridHeight)
    }

    companion object {

        internal const val RADIUS = 30
        internal const val GRID_WIDTH = 10
        internal const val GRID_HEIGHT = 10
        private const val GRID_X_FROM = 2
        private const val GRID_X_TO = 4
        private const val GRID_Z_FROM = 3
        private const val GRID_Z_TO = 5
        internal val ORIENTATION = HexagonOrientation.POINTY_TOP
    }
}
